tools/scylla-sstable: add lua scripting support (#11649)
scylladb/scylladb: docs/operating-scylla/admin-tools/scylla-sstable.rst

Parsing uses a streaming json parser, it is safe to pass in input-files of any size.

The output sstable will use the BIG format, the highest supported sstable format and the specified generation (``--generation``).
By default it is placed in the local directory, can be changed with ``--output-dir``.
If the output sstable clashes with an existing sstable, the write will fail.

The output is validated before being written to the disk.
The validation done here is similar to that done by the `validate operation <validate_>`_.
The level of validation can be changed with the ``--validation-level`` flag.
Possible validation-levels are:

* ``partition_region`` - only check fragment types, e.g. that a partition-end is followed by partition-start or EOS;
* ``token`` - also check token order of partitions;
* ``partition_key`` - full check on partition-ordering;
* ``clustering_key`` - also check clustering element ordering;

Note that levels are cumulative, each contains all the checks of the previous levels too.
By default the strictest level is used.
This can be relaxed if e.g. one wants to produce intentionally corrupt sstables for tests.

script
^^^^^^

.. versionadded:: 5.2

> Read the sstable(s) and pass the resulting `fragment stream <scylla-sstable-sstable-content_>`_. to the script specified by `--script-file`.
> Currently only script written in `Lua <http://www.lua.org/>`_ are supported.
> It is possible to pass command line arguments to the script, via the ``--script-arg`` command line flag.

Reviewer: ```suggestion
Reads the SStable(s) and passes the resulting `fragment stream <scylla-sstable-sstable-content_>`_ to the script specified by `--script-file`.
Currently, only scripts written in `Lua <http://www.lua.org/>`_ are supported.
It is possible to pass command line arguments to the script, via the ``--script-arg`` command line flag.
```

The format of this argument is the following:

.. code-block:: none
    :class: hide-copy-button

    --script-arg $key1=$value1:$key2=$value2

Alternatively, you can provide each key-value pair via a separate ``--script-arg``:

.. code-block:: none
    :class: hide-copy-button

    --script-arg $key1=$value1 --script-arg $key2=$value2

> Command line arguments will be receieved by the `consume_stream_start() <scylla-consume-stream-start-method_>`_ API method.

Reviewer: ```suggestion
Command line arguments will be received by the `consume_stream_start() <scylla-consume-stream-start-method_>`_ API method.
```

.. _scylla-consume-api:

> Scylla Consume API
> ~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
ScyllaDB Consume API
~~~~~~~~~~~~~~~~~~~~~~
```

These methods represent the glue code between scylla-sstable's C++ code and the Lua script.

> Conceptually a script is an implementation of a consumer interface. The script has to implement only the methods it is interested in, each method has a default implementation in the interface, which simply drops the respective `mutation fragment <scylla-sstable-sstable-content_>`_.
> For example a script only interested in partitions, can define only `consume_partition_start() <scylla-consume-partition-start-method_>`_ and nothing else.

Reviewer: ```suggestion
Conceptually a script is an implementation of a consumer interface. The script has to implement only the methods it is interested in. Each method has a default implementation in the interface, which simply drops the respective `mutation fragment <scylla-sstable-sstable-content_>`_.
For example, a script only interested in partitions can define only `consume_partition_start() <scylla-consume-partition-start-method_>`_ and nothing else.
```

> Conceptually a script is an implementation of a consumer interface. The script has to implement only the methods it is interested in, each method has a default implementation in the interface, which simply drops the respective `mutation fragment <scylla-sstable-sstable-content_>`_.
> For example a script only interested in partitions, can define only `consume_partition_start() <scylla-consume-partition-start-method_>`_ and nothing else.
> Therefore a completely empty script is also valid, although not very useful.
> Below you will find the listing of the API methods.
> These methods (if provided by the script) will be called by the scylla-sstable runtime for the appropriate events and fragment types.

Reviewer: ```suggestion
Conceptually, a script is an implementation of a consumer interface. The script has to implement only the methods it is interested in. Each method has a default implementation in the interface, which simply drops the respective `mutation fragment <scylla-sstable-sstable-content_>`_.
For example, a script only interested in partitions can define only `consume_partition_start() <scylla-consume-partition-start-method_>`_ and nothing else.
Therefore a completely empty script is also valid, although not very useful.
Below you will find the listing of the API methods.
These methods (if provided by the script) will be called by the scylla-sstable runtime for the appropriate events and fragment types.
```

.. _scylla-consume-stream-start-method:

consume_stream_start(args)
""""""""""""""""""""""""""

> Part of the Consume API. Called on the very start of the stream. Parameter is a lua table containing command line arguments for the script, passed via ``--script-arg``.
> Can be used to initialize global state.

Reviewer: ```suggestion
* Part of the Consume API. Called on the very start of the stream.
* Parameter is a Lua table containing command line arguments for the script, passed via ``--script-arg``.
* Can be used to initialize global state.
```

.. _scylla-consume-sstable-start-method:

consume_sstable_start(sst)
""""""""""""""""""""""""""

> Part of the Consume API. Called on the start of each sstable. Parameter is of type `Scylla.sstable <scylla-sstable-type_>`_. When sstables are merged (``--merge``) the parameter is ``nil``.
> Returns whether to stop. If ``true``, `consume_sstable_end() <scylla-consume-sstable-end-method_>`_ is called, skipping the content of the sstable (or that of the entire stream if ``--merge`` is used). If ``false``, consumption follows with the content of the sstable.

Reviewer: ```suggestion
* Part of the Consume API.
* Called on the start of each sstable. Parameter is of type `Scylla.sstable <scylla-sstable-type_>`_. 
* When SStables are merged (``--merge``), the parameter is ``nil``.
Returns whether to stop. If ``true``, `consume_sstable_end() <scylla-consume-sstable-end-method_>`_ is called, skipping the content of the sstable (or that of the entire stream if ``--merge`` is used). If ``false``, consumption follows with the content of the sstable.
```

.. _scylla-consume-partition-start-method:

consume_partition_start(ps)
"""""""""""""""""""""""""""

Part of the Consume API. Called on the start of each partition. Parameter is of type `Scylla.partition_start <scylla-partition-start-type_>`_.
Returns whether to stop. If ``true``, `consume_partition_end() <scylla-consume-partition-end-method_>`_ is called, skipping the content of the partition. If ``false``, consumption follows with the content of the partition.

consume_static_row(sr)
""""""""""""""""""""""

> Part of the Consume API. Called if the partition has a static row. Parameter is of type `Scylla.static_row <scylla-static-row-type_>`_.
> Returns whether to stop. If ``true``, `consume_partition_end() <scylla-consume-partition-end-method_>`_ is called, skipping the remaining content of the partition. If ``false``, consumption follows with the remaining content of the partition.

Reviewer: ```suggestion
* Part of the Consume API. 
* Called if the partition has a static row. 
* The parameter is of type `Scylla.static_row <scylla-static-row-type_>`_.
* Returns whether to stop. If ``true``, `consume_partition_end() <scylla-consume-partition-end-method_>`_ is called, and the remaining content of the partition is skipped. If ``false``, consumption follows with the remaining content of the partition.
```

consume_clustering_row(cr)
""""""""""""""""""""""""""

> Part of the Consume API. Called for each clustering row. Parameter is of type `Scylla.clustering_row <scylla-clustering-row-type_>`_.
> Returns whether to stop. If ``true``, `consume_partition_end() <scylla-consume-partition-end-method_>`_ is called, skipping the remaining content of the partition. If ``false``, consumption follows with the remaining content of the partition.

Reviewer: ```suggestion
* Part of the Consume API. 
* Called for each clustering row. 
* The parameter is of type `Scylla.clustering_row <scylla-clustering-row-type_>`_.
* Returns whether to stop. If ``true``, `consume_partition_end() <scylla-consume-partition-end-method_>`_ is called, the remaining content of the partition is skipped. If ``false``, consumption follows with the remaining content of the partition.
```

consume_range_tombstone_change(crt)
"""""""""""""""""""""""""""""""""""

Part of the Consume API. Called for each range tombstone change. Parameter is of type `Scylla.range_tombstone_change <scylla-range-tombstone-change-type_>`_.
Returns whether to stop. If ``true``, `consume_partition_end() <scylla-consume-partition-end-method_>`_ is called, skipping the remaining content of the partition. If ``false``, consumption follows with the remaining content of the partition.

.. _scylla-consume-partition-end-method:

consume_partition_end()
"""""""""""""""""""""""

> Part of the Consume API. Called at the end of the partition.
> Returns whether to stop. If ``true``, `consume_sstable_end() <scylla-consume-sstable-end-method_>`_ is called, skipping the remaining content of the sstable. If ``false``, consumption follows with the remaining content of the sstable.

Reviewer: ```suggestion
* Part of the Consume API.
* Called at the end of the partition.
* Returns whether to stop. If ``true``, `consume_sstable_end() <scylla-consume-sstable-end-method_>`_ is called,  the remaining content of the SStable is skipped. If ``false``, consumption follows with the remaining content of the SStable.
```

.. _scylla-consume-sstable-end-method:

consume_sstable_end()
"""""""""""""""""""""

Part of the Consume API. Called at the end of the sstable.
Returns whether to stop. If true, `consume_stream_end() <scylla-consume-stream-end-method_>`_ is called, skipping the remaining content of the stream. If false, consumption follows with the remaining content of the stream.

.. _scylla-consume-stream-end-method:

consume_stream_end()
""""""""""""""""""""

> Part of the Consume API. Called at the very end of the stream.

Reviewer: ```suggestion
* Part of the Consume API. 
* Called at the very end of the stream.
```

Scylla LUA API
~~~~~~~~~~~~~~

> In addition to the `Scylla Consume API <scylla-consume-api_>`_, the Lua bindings expose various types and methods to allow for working with Scylla types and values.

Reviewer: ```suggestion
In addition to the `ScyllaDB Consume API <scylla-consume-api_>`_, the Lua bindings expose various types and methods that allow you to work with ScyllaDB types and values.
```

> These are above those that the base, string and table Lua standard libraries expose.

Reviewer: > These are above those

@denesb 
I'm not sure what that means. This sentence needs some rewording, but I'm not sure about the meaning. I think it could be:
"These types and methods have priority over the types and methods exposed by the base, string, and table Lua standard libraries."
OR
"These types and methods should be preferred over the types and methods exposed by the base, string, and table Lua standard libraries."
